src/tree_annotation/database.cljs

```diff
@@ -25,172 +27,323 @@
+
+(defn stop-renaming [node]
+  (assoc node :renaming false))
+
+(defn stop-renaming-node! [index]
+  "Puts the node at `index` out of renaming mode."
+  (swap! db update :forest update-forest stop-renaming index))
+
+;; combine
+;; -------
+
+(defn combine [children]
+  "Combines the sequence `children` into a new tree."
+  (let [label (:label (last children))]
+    (assoc default-node
+           :label label
+           :children (mapv deselect-tree children)
+           :x (:x (first children))
+           :y (inc (reduce max (map :y children)))
+           :width (reduce + (map :width children)))))
+
+(defn combine-selected
+  ([forest]
+   "Finds groups of adjacent selected trees in `forest` and combines them."
+   (combine-selected forest [] []))
+  ([forest group acc]
+   (letfn [(combine-if [coll xs]
+             (if (empty? xs) coll (conj coll (combine xs))))]
+     (if (empty? forest)
+       ; end of forest
+       (combine-if acc group)
+       ; in forest
+       (let [tree (first forest)
+             tail (rest forest)]
+         (if (tree-selected? tree)
+           (combine-selected tail (conj group tree) acc)
+           (combine-selected tail [] (conj (combine-if acc group) tree))))))))
+
+(defn combine-selected! []
+  "Finds groups of adjacent selected trees and combines them."
+  (swap! db update :forest combine-selected))
+
+;; delete
+;; ------
+
+(defn node-delete-selected [node]
+  "Deletes all selected children of `node`.
+Returns either the unchanged node or a list of remaining subtrees."
+  (let [subtrees (mapv node-delete-selected (:children node))]
+    (if (or (and (:selected node) (not (leaf? node)))
+            (not-every? map? subtrees))
+      (vec (flatten subtrees))
+      node)))
+
+(defn delete-selected [forest]
+  "Delete all selected nodes and their ancestors from `forest`."
+  (vec (flatten (map node-delete-selected forest))))
+
+(defn delete-selected! []
+  "Delete all selected nodes and their ancestors."
+  (swap! db update :forest delete-selected))
+
+;; parse
+;; -----
+
+(declare recalc-coords)
+
+(defn recalc-coords-tree [node offset]
+  "Recalculate the coordinates in a tree with x offset `offset`."
+  (let [[children' offset-children] (recalc-coords (:children node) offset)
+        offset' (if (leaf? node) (inc offset-children) offset-children)
+        node' (assoc node
+                     :children children'
+                     :x offset
+                     :y (inc (reduce max 0 (map :y children')))
+                     :width (- offset' offset))]
+    [node' offset']))
+
+(defn recalc-coords [forest offset]
+  "Recalculates the coordinates in a `forest` with x offset `offset`."
+  (if (empty? forest)
+    [[] offset]
+    (let [tree (first forest)
+          tail (rest forest)
+          [tree' offset-children] (recalc-coords-tree tree offset)
+          [tail' offset-tail] (recalc-coords tail offset-children)]
+      [(into [tree'] tail') offset-tail])))
```

Comment:
I suggest to add a longer comment explaining the algorithm implemented by the next two functions and how it is used by the caller.

Reply:
This code becomes obsolete with the new layout (in branch `better-ui`), which doesn't rely on pre-computed coordinates but rather nests divs. This also allows fexible width nodes.